woltka/workflow.py

```diff
@@ -0,0 +1,590 @@
+        update_dict(tree, tree_)
+        update_dict(rankd, rankd_)
+
+    # Newick-format tree
+    if newick_fp:
+        with readzip(newick_fp) as f:
+            update_dict(tree, read_newick(f))
+
+    # lineage strings file
+    if lineage_fp:
+        with readzip(lineage_fp) as f:
+            tree_, rankd_ = read_lineage(f)
+        update_dict(tree, tree_)
+        update_dict(rankd, rankd_)
+
+    # rank table file
+    if ranktb_fp:
+        with readzip(ranktb_fp) as f:
+            update_dict(tree, read_ranktb(f))
+
+    # plain mapping files
+    for fp in map_fps:
+        rank = path2stem(fp)  # filename stem as rank
+        with readzip(fp) as f:
+            map_ = read_map(f)
+        update_dict(tree, map_)
+        if map_rank:
+            update_dict(rankd, {k: rank for k in set(map_.values())})
+
+    # fill root
+    root = fill_root(tree)
+
+    if is_build:
+        click.echo(' Done.')
+        click.echo(f'Total classification units: {len(tree)}.')
+    return tree, rankd, named, root
+
+
+def reshape_rmap(rmap:    dict,
+                 deidx:   bool = None,
+                 demux:   bool = None,
+                 samples: list = None,
+                 files:   dict = None,
+                 fp:      str = None) -> dict:
+    """Reshape a read map.
+
+    Parameters
+    ----------
+    rmap : dict
+        Read map to manipulate.
+    deidx : bool, optional
+        Strip suffixes from subject IDs.
+    demux : bool, optional
+        Whether perform demultiplexing.
+    samples : list of str, optional
+        Sample ID list to include.
+    files : list or dict
+        Map of filepaths to sample IDs.
+    fp : str
+        Path to current alignment file.
+
+    Returns
+    -------
+    dict
+        Reshaped read map.
+    """
+    # strip indices from subjects
+    if deidx:
+        strip_index(rmap)
+
+    # demultiplex into multiple samples
+    if demux:
+        return demultiplex(rmap, samples)
+
+    # sample Id from filename
+    else:
+        return {files[fp]: rmap}
+
+
+def assign_rmap(rmap:     dict,
+                data:     dict,
+                rank:     str,
+                sample:   str,
+                rank2dir: dict = None,
+                tree:     dict = None,
+                rankd:    dict = None,
+                named:    dict = None,
+                root:      str = None,
+                above:    bool = False,
+                major:     int = None,
+                ambig:     str = True,
+                subok:    bool = None):
+    """Assign query sequences in a query-to-subjects map to classification
+    units based on their subjects.
+
+    Parameters
+    ----------
+    rmap : dict
+        Read map to assign.
+    data : dict
+        Master data structure.
+    rank : str
+        Target rank to assign to.
+    sample : str
+        Sample ID.
+    rank2dir : dict, optional
+        Directory of output maps per rank.
+
```

Comment:
```suggestion
```

Reply:
For this question (whether empty line is fine in docstring) I am bit hesitating. There are too many parameters and some of them form groups. Strictly speaking they are all variables and have no hierarchies, but for better understandability it may make sense to separate them somehow. I just don't know how to add comments in docstring. Do you have suggestions? @gwarmstrong

Reply:
When documentation is concerned, I usually consult what has been done in `scikit-learn`, because I think they are really dedicated to making their documentation as neat and clear to the user as possible.

See: https://github.com/scikit-learn/scikit-learn/blob/95d4f0841/sklearn/neural_network/_multilayer_perceptron.py#L1083

It seems to me that what they do for a function with many parameters is not necessarily to group them. But instead give as much detailed information as possible.

Reply:
Thanks for pointing me to scikit-learn docstring. This is very helpful for me. They appeared to have an empty line for every parameter. Let me just remove empty lines.